vm/systemSmartContracts/auction.go

```diff
@@ -1191,6 +1360,99 @@ func (s *stakingAuctionSC) unBondTokens(args *vmcommon.ContractCallInput) vmcomm
 	return vmcommon.Ok
 }
 
+func (s *stakingAuctionSC) unBondNecessaryNodes(registrationData *AuctionDataV2, valueToUnBond *big.Int) (*big.Int, error) {
+	totalUnBond := big.NewInt(0)
+	for _, blsKey := range registrationData.BlsPubKeys {
+		if !s.isWithStatusInactive(blsKey) {
+			continue
+		}
+
+		nodeData, errGet := s.getStakedData(blsKey)
+		if errGet != nil {
+			return nil, errGet
+		}
+		if len(nodeData.RewardAddress) == 0 {
+			continue
+		}
+		// returns what value is still under the selected bls key
+		vmOutput, errExec := s.executeOnStakingSC([]byte("unBond@" + hex.EncodeToString(blsKey)))
+		if errExec != nil || vmOutput.ReturnCode != vmcommon.Ok {
+			continue
+		}
+
+		registrationData.NumRegistered -= 1
+		totalUnBond.Add(totalUnBond, s.baseConfig.NodePrice)
+		totalUnBond.Sub(totalUnBond, nodeData.SlashValue)
+
+		s.eei.Finish(blsKey)
+
+		if totalUnBond.Cmp(valueToUnBond) >= 0 {
+			return totalUnBond, nil
+		}
+	}
+
+	return totalUnBond, nil
+}
+
+func (s *stakingAuctionSC) unBondTokensWithNodes(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	registrationData, returnCode := s.basicCheckForUnStakeUnBond(args)
+	if returnCode != vmcommon.Ok {
+		return returnCode
+	}
+	err := s.eei.UseGas(s.gasCost.MetaChainSystemSCsCost.UnBondTokens)
+	if err != nil {
+		s.eei.AddReturnMessage(vm.InsufficientGasLimit)
+		return vmcommon.OutOfGas
+	}
+	if len(args.Arguments) != 1 {
+		s.eei.AddReturnMessage("invalid num of arguments")
+		return vmcommon.UserError
+	}
+
+	unBondValue := big.NewInt(0).SetBytes(args.Arguments[0])
+	if unBondValue.Cmp(zero) <= 0 {
+		s.eei.AddReturnMessage("cannot unStake negative value")
+		return vmcommon.UserError
+	}
+
+	totalUnBond, index := s.computeUnBondTokens(registrationData)
+	if totalUnBond.Cmp(zero) > 0 {
+		registrationData.UnstakedInfo = registrationData.UnstakedInfo[index:]
+		registrationData.TotalUnstaked.Sub(registrationData.TotalUnstaked, totalUnBond)
+		if registrationData.TotalUnstaked.Cmp(zero) < 0 {
+			s.eei.AddReturnMessage("contract error on unBondTokens function, total unstaked < total unbond")
+			return vmcommon.UserError
+		}
+	}
+
+	totalUnBondedFromNodes := big.NewInt(0)
+	if totalUnBond.Cmp(unBondValue) < 0 {
+		remainingToUnBond := big.NewInt(0).Sub(unBondValue, totalUnBond)
+		totalUnBondedFromNodes, err = s.unBondNecessaryNodes(registrationData, remainingToUnBond)
+		if err != nil {
+			s.eei.AddReturnMessage(err.Error())
+			return vmcommon.UserError
+		}
+	}
+
+	totalUnBond.Add(totalUnBond, totalUnBondedFromNodes)
+
+	s.eei.Finish(totalUnBond.Bytes())
+	err = s.eei.Transfer(args.CallerAddr, args.RecipientAddr, totalUnBond, nil, 0)
+	if err != nil {
+		s.eei.AddReturnMessage("transfer error on unBond function")
+		return vmcommon.UserError
+	}
```

Comment:
why not moving the rest of the funds (if there are any) in the TotalStakedValue?

Reply:
done